src/sources/socket/mod.rs

```diff
@@ -360,13 +444,84 @@ mod test {
         );
     }
 
+    #[test]
+    fn udp_shutdown_simple() {
+        let (tx, rx) = mpsc::channel(2);
+        let source_name = "udp_shutdown_simple";
+
+        let mut shutdown = ShutdownCoordinator::new();
+        let (address, mut rt, source_handle) =
+            init_udp_with_shutdown(tx, source_name, &mut shutdown);
+
+        send_lines_udp(address, vec!["test"]);
+        let events = rt.block_on(collect_n(rx, 1)).ok().unwrap();
+
+        assert_eq!(
+            events[0].as_log()[&event::log_schema().message_key()],
+            "test".into()
+        );
+
+        // Now signal to the Source to shut down.
+        shutdown.shutdown_source_begin(source_name);
+        let deadline = Instant::now() + Duration::from_secs(1000);
+        let shutdown_success =
+            shutdown.shutdown_source_end(&mut rt, source_name.to_string(), deadline);
+        assert_eq!(true, shutdown_success);
+
+        // Ensure source actually shut down successfully.
+        rt.block_on(source_handle).unwrap();
+    }
+
+    // #[test]
+    // fn udp_shutdown_infinite_stream() {
```

Comment:
So, this is the test I spent a lot of time on the last few days trying to get working.  I think I have finally concluded that it isn't possible with the version of the futures library we are using.  I left it here because I'd love to be wrong, so if someone knows how to get this working please let me know, I'll lay out the problem in another comment below.  Otherwise I'll plan on deleting this before pushing